proposals/text/6001-document-store-filter-rework.md

```diff
@@ -0,0 +1,197 @@
+
+As a practical example the below two filters are equivalent. Given that they're structured differently and `$and` is implicit in the second one:
+
+```
+{"number": {"$and": [{"$lte": 2}, {"$gte": 0}]}}
+
+{"number": {"$lte": 2, "$gte": 0}}
+```
+
+With the newly proposed approach the first filter would be equivalent to:
+
+```
+{
+    "operator": "AND",
+    "conditions": [
+        { "field": "number", "operator": "<=", "value": 2 },
+        { "field": "number", "operator": ">=", "value": 0 },
+    ]
+}
+```
+
+The second to:
+
+```
+{
+    "conditions": [
+        { "field": "number", "operator": "<=", "value": 2 },
+        { "field": "number", "operator": ">=", "value": 0 },
+    ]
+}
+```
+
+As you can see all the required information is one the same level and clearly recognisable. This makes it much easier both to read by a human and convert with code.
+
+In Python code:
+
+```python
+number <= 2 AND number >= 0
+```
+
+# Detailed design
+
+Filters top level must be a dictionary.
+
+There are two types of dictionaries:
+
+- Comparison
+- Logic
+
+Top level can be either be a Comparison or Logic dictionary.
+
+Comparison dictionaries must contain the keys:
+
+- `field`
+- `operator`
+- `value`
+
+Logic dictionaries have only one mandatory key:
+
+- `conditions`
+
+`conditions` key must be a list of dictionaries, either Comparison or Logic.
+
+The following key can be omitted:
+
+- `operator`
+
+`operator` values in Comparison dictionaries must be:
+
+- `==`
+- `!=`
+- `>`
+- `>=`
+- `<`
+- `<=`
+- `in`
+- `not in`
+
+`operator` values in Logic dictionaries must be:
+
+- `NOT`
+- `OR`
+- `AND`
+
+If `operator` key in Logic dictionaries is omitted we assume it's an `AND` value.
+
+---
+
+As an example this:
+
+```
+{
+    "$and": {
+        "type": {"$eq": "article"},
+        "$or": {"genre": {"$in": ["economy", "politics"]}, "publisher": {"$eq": "nytimes"}},
+        "date": {"$gte": "2015-01-01", "$lt": "2021-01-01"},
+        "rating": {"$gte": 3},
+    }
+}
+```
+
+would convert to:
+
+```
+{
+    "operator": "AND",
+    "conditions": [
+        { "field": "type", "operator": "==", "value": "article" },
+        {
+            "operator": "OR",
+            "conditions": [
+                { "field": "genre", "operator": "in", "value": ["economy", "politics"] },
+                { "field": "publisher", "operator": "==", "value": "nytimes" },
+            ]
+        },
+        { "field": "date", "operator": ">=", "value": "2015-01-01" },
+        { "field": "date", "operator": "<", "value": "2021-01-01" },
+        { "field": "rating", "operator": ">=", "value": 3 },
+    ]
+}
```

Comment:
What would it look like if I wanted to use the "NOT" Operator here to negate the entire filter? Could you please add an example where "NOT" is used?

Reply:
It would be the following right?
```
    "operator": "NOT",
    "conditions": [
            {
                "operator": "AND",
                "conditions": [
                        { "field":...
```

Reply:
Yep, that would be the case. It would be equivalent to this in the current implementation.

```
{
    "$not": {
        "whatever_field": ...
    }
}
```

It would work the same way as before, the same places you could use `$and` and `$or` you can use `$not`. With the new implementation it's the same, the same places you can use `AND` and `OR` you can use `NOT`.